[export] Support python assertion with symints.

zhxchen17 opened the pull request:
Summary: This diff ports some technique from torch.fx symbolic trace to trace through Python asserts when we run into data dependent symbolic shape assertions, so that we can achieve the same effect as torch dynamo to automatically turn assert into torch.check()s.

Test Plan: buck test mode/opt caffe2/test:test_export -- -r test_python_asserts_with_sym_int
Differential Revision: D71425360




cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @chenyang78 @kadeng @chauhang @amjames

tugsbayasgalan commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -6467,22 +6470,26 @@ def _check_frozen(self, expr: sympy.Basic, concrete_val: sympy.Basic) -> None:
         floc: Optional[Union[str, traceback.FrameSummary]] = framework_loc
         if floc is None:
-            frame = inspect.currentframe()
+            frame = self._get_user_frame()
```
Move this under try block?

zhxchen17 replied:
If this fails, finally block won't have `frame` variable to delete. There's no need to catch this directly.

tugsbayasgalan commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -6890,6 +6897,34 @@ def compute_concrete_val() -> sympy.Basic:
+                        assert frame is not None
+
+                        insts = list(dis.get_instructions(frame.f_code))
+                        if sys.version_info >= (3, 11):
```
Could you leave a comment explaining what this does?

tugsbayasgalan replied:
Maybe you can cross reference: https://github.com/pytorch/pytorch/blob/6bcf9c6ce349eeba4eeec19b40d6644b9ca7df2d/torch/_dynamo/symbolic_convert.py#L434 
In the dynamo code, they seem to talk about change in behaviour in 3.8, 3.9, 3.11, 3.12.

tugsbayasgalan commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -6890,6 +6897,34 @@ def compute_concrete_val() -> sympy.Basic:
+                            first = insts[cur + 1]
+
+                            starts_with_assert = (
+                                first.opname == "LOAD_GLOBAL"
```
Same here, can you cross reference with dynamo?

tugsbayasgalan commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -6890,6 +6897,34 @@ def compute_concrete_val() -> sympy.Basic:
+                            )
+                            if starts_with_assert and insts[cur + 2].opname == "RAISE_VARARGS":
+                                concrete_val = sympy.true
+                                transmute_into_runtime_assert = True
```
For my understanding: where does it actually add torch._check in symbolic tracing? I thought you were supposed to do sth like https://github.com/pytorch/pytorch/pull/120564/files#diff-622913fdb49db90d6f3a8ab225b4badb7996023e6498e9f7c6d03fe9f32d0986R351

zhxchen17 replied:
I think there might be recursion issue there since we are already lower in the stack here, we can do that as BE later.

tugsbayasgalan commented on test/export/test_export.py:
```diff
@@ -10086,6 +10086,24 @@ def forward(self, x, b_t, y):
+                return y
+
+        model = Model()
+        ep = torch.export.export(model, (torch.zeros(4, dtype=torch.int),))
```
How does the graph look like?

pianpwk commented on torch/fx/experimental/symbolic_shapes.py:
```diff
@@ -6890,6 +6897,34 @@ def compute_concrete_val() -> sympy.Basic:
                         concrete_val = unsound_result
                         ok = True
 
+                    if not ok and self.trace_asserts:
```
this feels pretty cool, it's basically interpreting asserts as calls to expect_true(). Could we maybe name it something more informative and intimidating like "defer_asserts_as_runtime_asserts" (terrible name lol)? I feel like currently it looks too friendly, everyone will want to turn it on.

We could also move this behavior under `prefer_deferred_runtime_asserts_over_guards`, that's super similar (turns expect_trues into runtime asserts) and I believe export is the only user of this flag.

nit: I don't love where this code is, since it feels a bit orthogonal to symbolic shapes, but am not sure what a better place is. Maybe we could catch it earlier in SymNode.guard_bool(), so we just redirect to expect_true() immediately? I feel like that also has the nice predictability of "every assert gets turned into a runtime assert, not just the ones that would've failed on data-dependent": https://github.com/pytorch/pytorch/blob/981807cfcbc2dc8036f0e83f038455b49ba4b833/torch/fx/experimental/sym_node.py#L533

Also, does this work for both strict & non-strict?

tugsbayasgalan replied:
I think dynamo already does it for strict

zhxchen17 replied:
yeah we already have this for strict. This is just fixing the behavior on non-strict side to bring two closer.

zhxchen17 commented:
added more docs about the behavior of bytecode

zhxchen17 committed:
[export] Support python assertion with symints. (#149444)

Summary:

This diff ports some technique from torch.fx symbolic trace to trace through Python asserts when we run into data dependent symbolic shape assertions, so that we can achieve the same effect as torch dynamo to automatically turn assert into torch.check()s.

Test Plan: buck test mode/opt caffe2/test:test_export -- -r test_python_asserts_with_sym_int

Differential Revision: D71425360
